contracts/Compass.vy

```diff
@@ -223,12 +320,175 @@ def submit_batch(consensus: Consensus, token: address, args: TokenSendArgs, batc
+            cumulative_power += consensus.valset.powers[i]
+            if cumulative_power >= POWER_THRESHOLD:
+                break
+        i = unsafe_add(i, 1)
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+
+# Make a new checkpoint from the supplied validator set
+# A checkpoint is a hash of all relevant information about the valset. This is stored by the contract,
+# instead of storing the information directly. This saves on storage and gas.
+# The format of the checkpoint is:
+# keccak256 hash of abi_encoded checkpoint(validators[], powers[], valset_id, compass_id)
+# The validator powers must be decreasing or equal. This is important for checking the signatures on the
+# next valset, since it allows the caller to stop verifying signatures once a quorum of signatures have been verified.
+@internal
+@pure
+def make_checkpoint(valset: Valset) -> bytes32:
+    return keccak256(_abi_encode(valset.validators, valset.powers, valset.valset_id, compass_id, method_id=method_id("checkpoint(address[],uint256[],uint256,bytes32)")))
+
+@internal
+def transfer_fees(fee_args: FeeArgs, total:uint256):
+    self.claimable_rewards[fee_args.relayer] = unsafe_add(self.claimable_rewards[fee_args.relayer], fee_args.relayer_fee)
+    self.rewards_community_wallet = unsafe_add(self.rewards_community_wallet, fee_args.community_fee)
+    self.rewards_security_wallet = unsafe_add(self.rewards_security_wallet, fee_args.security_fee)
+    self.funds[fee_args.fee_payer_paloma_address] = unsafe_sub(self.funds[fee_args.fee_payer_paloma_address], total)
+    self.total_claims = unsafe_add(self.total_claims, fee_args.relayer_fee)
+    self.total_funds = unsafe_sub(self.total_funds, total)
+
+@internal
+@view
+def assert_fees(fee_args: FeeArgs) -> uint256:
+    _total_fee:uint256 = fee_args.relayer_fee + fee_args.community_fee + fee_args.security_fee
+    assert self.funds[fee_args.fee_payer_paloma_address] >= _total_fee, "Insufficient user funds to cover fees"
+    return _total_fee
+
+@internal
+@pure
+def assert_authority(authority: address, gas_estimate: uint256):
+    assert authority == msg.sender, "Message sender unauthorized."
+    # aviod reimbursing more than relayer was willing to pay
+    assert msg.gas >= gas_estimate, "Insufficient funds to cover gas estimate"
+
+@internal
+def reserve_security_refund(sender: address, gas_estimate: uint256):
+    # Refund is only covered while security wallet is funded.
+    if self.rewards_security_wallet >= gas_estimate:
+        self.rewards_security_wallet = unsafe_sub(self.rewards_security_wallet, gas_estimate)
+        self.claimable_rewards[sender] = unsafe_add(self.claimable_rewards[sender], gas_estimate)
+        self.total_claims = unsafe_add(self.total_claims, gas_estimate)
+
+@internal
+@view
+def assert_balance():
+    assert self.balance == self.total_claims + self.total_funds + self.rewards_security_wallet + self.rewards_community_wallet, "balance mismatch"
```

Comment:
This can't work. If someone sends some funds to compass-evm, the compass will die.

Reply:
Good point. I'm guessing there's no way to prevent random funds being sent to the address.

This is from my days of working with RDBS. I'm not sure whether such an assertion would even be needed. But in principle, we can replace the `==` with an `>=` and we're still in the safe zone.